### PR TITLE
displif v1

### PR DESCRIPTION
**Pull request by andr2000**
Please review

### CODE REVIEW

File: xen/include/public/io/displif.h [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+
+/*
+ * EVENT CODES.
+ */
+#define XENDISPL_EVT_PG_FLIP              0
+
+/*
+ * XENSTORE FIELD AND PATH NAME STRINGS, HELPERS.
+ */
+#define XENDISPL_DRIVER_NAME                   "vdispl"
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Why different indent?

File: xen/include/public/io/displif.h [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+    uint32_t padding;
+    uint64_t fb_cookie;
+};
+
+union xendispl_req {
+    struct {
+        uint16_t id;
+        uint8_t operation;
+        uint8_t reserved;
+        uint32_t·padding;
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Tab here?

File: xen/include/public/io/displif.h [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+        int8_t status;
+    } data;
+    uint8_t raw[64];
+};
+
+union xendispl_evt {
+    struct {
+        uint16_t id;
+        uint8_t type;
+        uint8_t·padding[5];
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Same here

File: xen/include/public/io/displif.h [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+};
+
+#define XENDISPL_IN_RING_OFFS (sizeof(union xendispl_event_page))
+#define XENDISPL_IN_RING_SIZE (PAGE_SIZE - XENDISPL_IN_RING_OFFS)
+#define XENDISPL_IN_RING_LEN (XENDISPL_IN_RING_SIZE / sizeof(union xendispl_evt))
+#define XENDISPL_IN_RING(page) \
+	((union xendispl_evt *)((char *)(page) + XENDISPL_IN_RING_OFFS))
+#define XENDISPL_IN_RING_REF(page, idx) \
+	(XENDISPL_IN_RING((page))[(idx) % XENDISPL_IN_RING_LEN])
+
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Any comments for these? I haven't seen them before in pv sound

File: xen/include/public/io/displif.h [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+ * operation - uint8_t, XENDISPL_OP_PG_FLIP
+ * crtc_idx - uint32_t, private guest value, echoed in response by the back
+ * fb_cookie - uint64_t, unique to guest domain value used by the backend
+ *   to map remote framebuffer to local in requests
+ */
+
+struct xendispl_page_flip_req {
+    uint32_t crtc_idx;
+    uint32_t padding;
+    uint64_t fb_cookie;
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Sometimes cookie in the beginning of struct, sometimes in the eo struct...

File: xen/include/public/io/displif.h [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+ * |                               reserved                                |
+ * +-----------------+-----------------+-----------------+-----------------+
+ * |/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/|
+ * +-----------------+-----------------+-----------------+-----------------+
+ * |                               reserved                                |
+ * +-----------------+-----------------+-----------------+-----------------+
+ *
+ * id - uint16_t, private guest value, echoed in response
+ * operation - uint8_t, XENDISPL_OP_FB_DESTROY
+ * dumb_cookie - uint64_t, unique to guest domain value used by the backend
```
Comment (klogg) [inline xen/include/public/io/displif.h]: inconsistent with structure contents

File: xen/include/public/io/displif.h [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+ *-------------------------------- Addressing ---------------------------------
+ *
+ * Indices used to address frontends, driver instances, cards,
+ * devices and streams.
+ *
+ * frontend_id
+ *      Values:         <uint>
+ *
+ *      Domain ID of the sound frontend.
+ *
```
Comment (klogg) [inline xen/include/public/io/displif.h]: sound?

File: xen/include/public/io/displif.h [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+ * STATE DIAGRAMS
+ *
+ *****************************************************************************
+ *                                   Startup                                 *
+ *****************************************************************************
+ *
+ * Tool stack creates front and back state nodes with initial state
+ * XenbusStateInitialising.
+ * Tool stack creates and sets up frontend sound configuration nodes per domain.
+ *
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Sound?

File: xen/include/public/io/displif.h [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+ *              |
+ *              |
+ *              V
+ * XenbusStateConnected
+ *
+ *                                      XenbusStateUnknown
+ *                                      XenbusStateClosed
+ *                                      XenbusStateClosing
+ * o Remove virtual sound device
+ * o Remove event channels
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Sound?

File: xen/include/public/io/displif.h [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+ * Shared page contains a ring with request/response packets.
+ *
+ *****************************************************************************
+ *                            Frontend to backend requests
+ *****************************************************************************
+ *
+ * All request packets have the same length (64 octets)
+ *
+ *
+ * Request dumb creation - request creation of a DISPL dumb buffer.
```
Comment (klogg) [inline xen/include/public/io/displif.h]: Probably you need to explain somewhere in comments decisions on memory sharing and buffers (dumb, dma, tilers). Also you need to exlpain set of currently supported features (i.e. do you plan to add planes? renderers? etc.)

File: xen/include/public/io/displif.h (hunk 1 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 2 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 3 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 4 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: as fb and kbd - will think about

File: xen/include/public/io/displif.h (hunk 5 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 6 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 7 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h (hunk 9 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+#define XENDISPL_RSP_OKAY                 0
+
+/* Path entries */
+#define XENDISPL_PATH_CONNECTOR           "connector"
+
+/*
+ * Description of the protocol between frontend and backend driver.
+ *
+ * The two halves of a Para-virtual sound driver communicates with
+ * each other using a shared page and an event channel.
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: done

**Comment by andr2000** [comment]
please review the updated version

File: xen/include/public/io/displif.h [hunk 12]
```diff
@@ -0,0 +1,620 @@
...
+    uint8_t padding[60];
+};
+
+struct xendispl_evt {
+    uint16_t id;
+    uint8_t type;
+    uint8_t reserved[5];
+    union {
+        struct xendispl_pg_flip_evt pg_flip;
+        uint8_t raw[56];
```
Comment (klogg) [inline xen/include/public/io/displif.h]: In the xendispl_req you have padding[], here you have raw[]... which one is correct?

File: xen/include/public/io/displif.h [hunk 13]
```diff
@@ -0,0 +1,620 @@
+/******************************************************************************
+ * displif.h
+ *
+ * Unified display device I/O interface for Xen guest OSes.
```
Comment (klogg) [inline xen/include/public/io/displif.h]: I guess this needs to be extended to better explain what this driver is and, more importantly, what it is not...

File: xen/include/public/io/displif.h [hunk 14]
```diff
@@ -0,0 +1,620 @@
...
+ * Front->back notifications: When enqueuing a new request, sending a
+ * notification can be made conditional on req_event (i.e., the generic
+ * hold-off mechanism provided by the ring macros). Backends must set
+ * req_event appropriately (e.g., using RING_FINAL_CHECK_FOR_REQUESTS()).
+ *
+ * Back->front notifications: When enqueuing a new response, sending a
+ * notification can be made conditional on rsp_event (i.e., the generic
+ * hold-off mechanism provided by the ring macros). Frontends must set
+ * rsp_event appropriately (e.g., using RING_FINAL_CHECK_FOR_RESPONSES()).
+ */
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: this is the part of generic description found in all protocols. won't fix

File: xen/include/public/io/displif.h (hunk 13 above)
Comment (andr2000) [inline xen/include/public/io/displif.h]: Well, started to describe and it comes out that the protocol is mostly self-explanatory.
Focused more on TODO part to better demonstrate what this protocol can be

Comment (andr2000) [inline xen/include/public/io/displif.h]: done

File: xen/include/public/io/displif.h [hunk 15]
```diff
@@ -0,0 +1,620 @@
...
+ * fb_cookie - uint64_t, unique to guest domain value used by the backend
+ *   to map remote framebuffer to local in requests
+ *
+ */
+
+struct xendispl_pg_flip_evt {
+    uint64_t fb_cookie;
+    uint32_t conn_idx;
+    uint32_t padding;
+};
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: removed

File: xen/include/public/io/displif.h [hunk 16]
```diff
@@ -0,0 +1,620 @@
...
+        struct xendispl_fb_create_req fb_create;
+        struct xendispl_fb_destroy_req fb_destroy;
+        struct xendispl_set_config_req set_config;
+        struct xendispl_page_flip_req pg_flip;
+        uint8_t padding[56];
+    } op;
+};
+
+struct xendispl_resp {
+    uint16_t id;
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: why?

File: xen/include/public/io/displif.h [hunk 17]
```diff
@@ -0,0 +1,620 @@
...
+    uint8_t type;
+    uint8_t reserved[5];
+    union {
+        struct xendispl_pg_flip_evt pg_flip;
+        uint8_t raw[56];
+    } op;
+};
+
+DEFINE_RING_TYPES(xen_displif, struct xendispl_req, struct xendispl_resp);
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: no, it is correct

Comment (andr2000) [inline xen/include/public/io/displif.h]: events are delivered via their own ring buffer which is different from req/resp ring

File: xen/include/public/io/displif.h [hunk 18]
```diff
@@ -0,0 +1,620 @@
...
+
+union xendispl_event_page {
+    struct {
+        uint32_t in_cons;
+        uint32_t in_prod;
+    } ring;
+    uint8_t raw[32];
+};
+
+#define XENDISPL_IN_RING_OFFS (sizeof(struct xendispl_event_page))
```
Comment (andr2000) [inline xen/include/public/io/displif.h]: will do

**Pull request closed without merging**